### PR TITLE
drop lsan suppressions

### PR DESCRIPTION
**Pull request by nshy**
Required small changes  - https://github.com/tarantool/small/pull/91
EE part - https://github.com/tarantool/tarantool-ee/pull/835

### CODE REVIEW

File: changelogs/unreleased/fix-memory-leak-in-iconv.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+## bugfix/lua
+
+- Fixed iconv object memory leak.
```
Comment (p7nov) [inline changelogs/unreleased/fix-memory-leak-in-iconv.md:3]:
```suggestion
- Fixed an iconv object memory leak.
```

File: changelogs/unreleased/fix-memory-leak-in-vinyl-on-dump-failure.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+## bugfix/vinyl
+
+- Fixed memory leak in vinyl on dump/compaction failure.
```
Comment (p7nov) [inline changelogs/unreleased/fix-memory-leak-in-vinyl-on-dump-failure.md:3]:
```suggestion
- Fixed a memory leak in vinyl on dump/compaction failure.
```

File: changelogs/unreleased/fix-misc-memory-leaks.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
+## bugfix/box
+
+- Fixed memory leak on disconnectiong from replica.
+- Fixed memory leak on user ddl on access check failure.
+- Fixed memory leak on xlog open failure.
+- Fixed memory leak on space ddl on certain condition.
+- Fixed memory leak on iproto override usage (since 3.1).
+- Fixed memory leak on ddl failed due to foreign key constraint check.
+- Fixed memory leak on space drop with sequence bound to json path.
```
Comment (p7nov) [inline changelogs/unreleased/fix-misc-memory-leaks.md:9]:
```suggestion
- Fixed a memory leak on disconnection from a replica.
- Fixed a memory leak on user DDL on access check failure.
- Fixed a memory leak on `xlog` open failure.
- Fixed a memory leak on space ddl on certain condition.
- Fixed a memory leak on iproto override usage (since 3.1).
- Fixed a memory leak on DDL failed due to foreign key constraint check.
- Fixed a memory leak on space drop with sequence bound to json path.
```

**Review by p7nov — APPROVED** [review]

File: src/box/alter.cc [hunk 4]
```diff
@@ -1021,7 +1021,9 @@ alter_space_do(struct txn_stmt *stmt, struct alter_space *alter)
 
 	alter->new_space->sequence = alter->old_space->sequence;
 	alter->new_space->sequence_fieldno = alter->old_space->sequence_fieldno;
-	alter->new_space->sequence_path = alter->old_space->sequence_path;
+	if (alter->old_space->sequence_path != NULL)
+		alter->new_space->sequence_path =
+			(char *)xstrdup(alter->old_space->sequence_path);
```
Comment (locker) [inline src/box/alter.cc:1026]: Let's split this patch into different commits, for example:
- Leak on DDL access check failure.
- Leak on xlog open failure.
- Leak on iproto override.
- Leak on replication reconfiguration.
- Leak on network address resolution error.

Also, it's preferable to submit some fixes in separate PRs because a PR with > 5-7 commits is difficult to review.

File: src/lib/swim/swim.c [hunk 5]
```diff
@@ -2252,6 +2252,10 @@ swim_delete(struct swim *swim)
 void
 swim_gc(struct swim *swim)
 {
+	if (swim->event_handler == NULL) {
```
Comment (locker) [inline src/lib/swim/swim.c:2255]: Please reference the commit that introduced the leak in the commit message.

File: src/box/lua/space.cc [hunk 6]
```diff
@@ -765,12 +765,16 @@ box_lua_space_delete(struct lua_State *L, uint32_t id)
...
+	(void)trigger;
+	struct lua_State *L = tarantool_L;
 	struct space *space = (struct space *) event;
 
 	if (space_by_id(space->def->id) != NULL) {
 		box_lua_space_new(L, space);
 	} else {
+		/* tarantool_L is freed on Tarantool shutdown. */
+		if (L == NULL)
+			return 0;
```
Comment (locker) [inline src/box/lua/space.cc:777]: Shouldn't box be destroyed before Lua? I don't like that there are so many places now where we have to check `tarantool_L`.

File: src/box/memtx_engine.cc [hunk 7]
```diff
@@ -1774,7 +1780,15 @@ memtx_engine_schedule_gc(struct memtx_engine *memtx,
 			 struct memtx_gc_task *task)
 {
 	stailq_add_tail_entry(&memtx->gc_queue, task, link);
-	fiber_wakeup(memtx->gc_fiber);
+	if (memtx->gc_fiber == NULL) {
+#ifdef ENABLE_ASAN
+		bool stop = false;
+		while (!stop)
+			memtx_engine_run_gc(memtx, &stop);
```
Comment (locker) [inline src/box/memtx_engine.cc:1787]: I think this should be done when the engine is destroyed, not when we schedule garbage collection.

File: src/box/txn_limbo.h [hunk 8]
```diff
@@ -480,6 +480,12 @@ txn_limbo_unfence(struct txn_limbo *limbo);
 void
 txn_limbo_init();
 
+/**
+ * Denitialize qsync engine.
+ */
+void
+txn_limbo_destroy();
```
Comment (locker) [inline src/box/txn_limbo.h:487]: Should be named `txn_limbo_free` according to our naming convention. Anyway, there are so many susbystems touched in this commit with so different cases so it may be better to split this commit as well and submit it in a different PR.

File: asan/lsan.supp [hunk 9]
```diff
@@ -4,77 +4,9 @@
...
-leak:Exception::operator new
-
-# test: vinyl/errinj.test.lua
-# source: src/lib/core/fiber.h
-leak:fiber_cxx_invoke
-
-# TODO(gh-9213): remove when the issue is fixed
-leak:luaT_push_key_def
-
-# TODO(gh-8890): remove when the issue is fixed
```
Comment (locker) [inline asan/lsan.supp:75]: Let's add `Closes #XXXX` to commit messages where applicable. Also, it'd be better if you removed the suppressions in the commits that actually fix the leaks.

**Comment by nshy** [comment]
Closed as superseded by #10475 #10491 #10500 #10505 #10516.
